src/lib/y2users/users_module/reader.rb

```diff
@@ -0,0 +1,173 @@
+      # @return [User]
+      def user(user_attrs)
+        user = User.new(user_attrs["uid"])
+
+        user.uid   = attr_value(:uidNumber, user_attrs)
+        user.gid   = attr_value(:gidNumber, user_attrs)
+        user.shell = attr_value(:loginShell, user_attrs)
+        user.home  = attr_value(:homeDirectory, user_attrs)
+        user.gecos = [attr_value(:cn, user_attrs), *user_attrs["addit_data"].split(",")].compact
+        if !user.uid && user_attrs["type"] == "system"
+          user.system = true
+        end
+
+        # set password only if specified, nil means not touch it
+        user.password = create_password(user_attrs) if user_attrs["userPassword"]
+
+        user
+      end
+
+      # Creates a {Group} object based on the data structure of a Users user
+      #
+      # @param group_attrs [Hash] a group representation in the format used by Users
+      # @return [Group]
+      def group(group_attrs)
+        group = Group.new(group_attrs["cn"])
+
+        group.gid   = attr_value(:gidNumber, group_attrs)
+        group.users_name = group_attrs["userlist"].keys
+
+        # TODO: no system support for groups
+        # if !group.gid && group_attrs["type"] == "system"
+        #  group.system = true
+        # end
+
+        # set password only if specified, nil means not touch it
+        # TODO: group passwords not supported
+        # group.password = create_password(group_attrs) if group_attrs["userPassword"]
+
+        group
+      end
+
+      # Value of the given field for the given Users user
+      #
+      # @param attr [#to_s] name of the attribute
+      # @param user [Hash] a user representation in the format used by Users
+      # @return [String, nil] nil if the value is missing or empty
+      def attr_value(attr, user)
+        value = user[attr.to_s]
+        return nil if value == ""
+
+        value
+      end
+
+      # Creates a {Password} object based on the data structure of a UsersSimple user
+      #
+      # @param user [Hash] a user representation in the format used by UsersSimple
+      # @return [Password]
+      def create_password(user)
+        create_method = user["encrypted"] ? :create_encrypted : :create_plain
+
+        Password.public_send(create_method, user["userPassword"]).tap do |password|
+          last_change = user["shadowLastChange"]
+          expiration = user["shadowExpire"]
+
+          password.aging = PasswordAging.new(last_change) if last_change
+          password.minimum_age = user["shadowMin"]
+          password.maximum_age = user["shadowMax"]
+          password.warning_period = user["shadowWarning"]
+          password.inactivity_period = user["shadowInactive"]
+          password.account_expiration = AccountExpiration.new(expiration) if expiration
+        end
+      end
+    end
+  end
+end
+
+if $0 == __FILE__
+  require "pp"
+
+  Yast::Users.Read
+  pp Y2Users::UsersModule::Reader.new.read
+end
```

Comment:
NP: to be removed :wink:

Reply:
Ah, it's in your check list in the description

> remove testing reader run

I overlooked it

Reply:
yep, it is in todo. But very useful for quick testing